crates/transaction-pool/src/pool/parked.rs

```diff
@@ -79,10 +87,106 @@ impl<T: ParkedOrd> ParkedPool<T> {
         Some(tx.transaction.into())
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.best.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Get transactions by sender
+    pub(crate) fn get_txs_by_sender(&self, sender: SenderId) -> Vec<TransactionId> {
+        self.by_id
+            .range((sender.start_bound(), Unbounded))
+            .take_while(move |(other, _)| sender == other.sender)
+            .map(|(_, tx)| *tx.transaction.id())
+            .collect()
+    }
+
+    /// Returns sender ids sorted by each sender's last submission id. Senders with older last
+    /// submission ids are first. Note that _last_ submission ids are the newest submission id for
+    /// that sender, so this sorts senders by the last time they submitted a transaction in
+    /// descending order. Senders that have least recently submitted a transaction are first.
+    ///
+    /// Similar to `Heartbeat` in Geth
+    pub fn get_senders_by_submission_id(&self) -> Vec<SubmissionSenderId> {
+        // iterate through by_id, and get the last submission id for each sender
+        let senders = self
+            .by_id
+            .iter()
+            .fold(Vec::new(), |mut set: Vec<SubmissionSenderId>, (_, tx)| {
+                if let Some(last) = set.last_mut() {
+                    // sort by last
+                    if last.sender_id == tx.transaction.sender_id() {
+                        if last.submission_id < tx.submission_id {
+                            // update last submission id
+                            last.submission_id = tx.submission_id;
+                        }
+                    } else {
+                        // new entry
+                        set.push(SubmissionSenderId::new(
+                            tx.transaction.sender_id(),
+                            tx.submission_id,
+                        ));
+                    }
+                } else {
+                    // first entry
+                    set.push(SubmissionSenderId::new(tx.transaction.sender_id(), tx.submission_id));
+                }
+                set
+            })
+            .into_iter()
+            // sort by submission id
+            .collect::<BinaryHeap<_>>();
```

Comment:
not exactly sure how this works internally but I assume this into_iter collect is optimized so there are no allocs,

but I think this could be written as a loop over an iterator that pushes unique senders directly into a binaryheap